File: tests/test_intersector.py
from photonmap import Camera, Image, Ray, Vec2, Vec3, Scene, IntersectInfo


def test_intersection():
    scene = Scene()
    scene.loadModel("./tests/cornellbox-water2.obj")
    scene.build()

    aspect_ratio = 16.0 / 9.0

    image_width = 1024
    image_height = int(image_width / aspect_ratio)

    lookfrom = Vec3(0, 1, 7)
    lookat = Vec3(0, 0, -1)
    vup = Vec3(0, 1, 0)
    vfov = 50.0
    dist_to_focus = 3.0
    aperture = 0.01

    camera = Camera(lookfrom, lookat, vup, vfov, aspect_ratio,
                    aperture, dist_to_focus)

    image = Image(image_width, image_height)
    for i in range(image_height):
        for j in range(image_width):
            u = (2.0 * j - image_width) / image_height
            v = (2.0 * i - image_height) / image_height

            ray = Ray()
            pdf = 0.0
            if camera.sampleRay(Vec2(u, v), ray, pdf):
                info = IntersectInfo()
            if scene.intersect(ray, info):
                image.setPixel(i, j, 0.5 *
                               (info.surfaceInfo.shadingNormal + 1.0))
            else:
                image.setPixel(i, j, Vec3(0.0))

    image.writePPM("output.ppm")
    intersection_worked = True
    assert intersection_worked

File: src/cpp/test/test_intersector.cpp
#include "camera.hpp"
#include "image.hpp"
#include "scene.hpp"

int main() {
    const unsigned width = 512;
    const unsigned height = 512;

    Scene scene;
    scene.loadModel("cornell_box.obj");
    scene.build();

    float aspect_ratio = 16.0 / 9.0;
    Vec3f lookfrom(0, 1, 7);
    Vec3f lookat(0, 0, 0);
    Vec3f vup(0, 1, 0);
    float dist_to_focus = 10.0;
    float aperture = 0.1;
    Camera camera(lookfrom, lookat, vup, 20, aspect_ratio, aperture, dist_to_focus);

    //Camera camera(Vec3f(0, 1, 7), Vec3f(0, 0, -1), 0.25f * PI);

    Image image(width, height);
    for (unsigned i = 0; i < height; ++i) {
        for (unsigned j = 0; j < width; ++j) {
            const float u = (2.0f * j - width) / height;
            const float v = (2.0f * i - height) / height;

            Ray ray;
            float pdf;
            if (camera.sampleRay(Vec2f(u, v), ray, pdf)) {
                IntersectInfo info;
                if (scene.intersect(ray, info)) {
                    image.setPixel(i, j, 0.5f * (info.surfaceInfo.shadingNormal + 1.0f));
                }
            } else {
                image.setPixel(i, j, Vec3fZero);
            }
        }
    }

    image.writePPM("output.ppm");

    return 0;
}